services/synccommitteemessenger/standard/service.go

```diff
@@ -176,47 +176,44 @@ func (s *Service) Message(ctx context.Context, duty *synccommitteemessenger.Duty
 
 	// Sign in parallel.
 	msgs := make([]*altair.SyncCommitteeMessage, 0, len(duty.ContributionIndices()))
-	var msgsMu sync.Mutex
 	validatorIndices := duty.ValidatorIndices()
 
 	s.UpdateSyncCommitteeDataRecord(duty.Slot(), *beaconBlockRoot, duty.ContributionIndices())
 
-	var wg sync.WaitGroup
+	accounts := make([]e2wtypes.Account, len(validatorIndices))
+
 	for i := range validatorIndices {
-		wg.Add(1)
-		go func(ctx context.Context,
-			wg *sync.WaitGroup,
-			i int,
-		) {
-			defer wg.Done()
-			account := duty.Account(validatorIndices[i])
-			if account == nil {
-				s.log.Debug().Msg("Account nil; likely exited validator still in sync committee")
-				return
-			}
-			sig, err := s.contribute(ctx, account, s.chainTimeService.SlotToEpoch(duty.Slot()), *beaconBlockRoot)
-			if err != nil {
-				s.log.Error().Err(err).Msg("Failed to sign sync committee message")
-				return
-			}
-			s.log.Trace().
-				Uint64("slot", uint64(duty.Slot())).
-				Uint64("validator_index", uint64(validatorIndices[i])).
-				Stringer("signature", sig).
-				Msg("Signed sync committee message")
-
-			msg := &altair.SyncCommitteeMessage{
-				Slot:            duty.Slot(),
-				BeaconBlockRoot: *beaconBlockRoot,
-				ValidatorIndex:  validatorIndices[i],
-				Signature:       sig,
-			}
-			msgsMu.Lock()
-			msgs = append(msgs, msg)
-			msgsMu.Unlock()
-		}(ctx, &wg, i)
+		account := duty.Account(validatorIndices[i])
+		if account == nil {
+			s.log.Debug().Msg("Account nil; likely exited validator still in sync committee")
+			continue
+		}
+		accounts[i] = account
+	}
+	sigs, err := s.contributions(ctx, accounts, s.chainTimeService.SlotToEpoch(duty.Slot()), *beaconBlockRoot)
+	if err != nil {
+		s.log.Error().Err(err).Msg("Failed to sign sync committee messages")
+		return nil, errors.Wrap(err, "failed to sign sync committee messages")
+	}
+
+	for i := range accounts {
+		if len(sigs) == 0 || len(sigs) < i {
```

Comment:
It looks like this length check doesn't need to be in the loop, it could sit after fetching `sigs` and be executed only once.

Reply:
`sigs` is a fixed size array so this check didn't really make sense anyway. I have replaced with a zero check on the sig for active accounts.